packages/acceptance-tests/pkg-tests-specs/sources/pnpapi.test.js

```diff
@@ -24,6 +24,33 @@ describe(`Plug'n'Play API`, () => {
       }),
     );
 
+    test(
+      `it shouldn't mess up when using createRequire on virtual files`,
+      makeTemporaryEnv({
+        private: true,
+        workspaces: [
+          `workspace`,
+        ],
+        dependencies: {
+          [`workspace`]: `workspace:*`,
+          [`no-deps`]: `1.0.0`,
+        },
+      }, async ({path, run, source}) => {
+        await xfs.mkdirpPromise(`${path}/workspace`);
+        await xfs.writeJsonPromise(`${path}/workspace/package.json`, {
+          name: `workspace`,
+          peerDependencies: {
+            [`no-deps`]: `*`,
+          },
+        });
+
+        await run(`install`);
+
+        const resolution = await source(`require('module').createRequire(require.resolve('workspace/package.json')).resolve('no-deps')`);
+        expect(resolution.match(/\.yarn/g)).toHaveLength(1);
```

Comment:
I'm not sure happy of this test (I'd have preferred to check whether requiring a module works), but I didn't find immediately the exact circumstances under which the require would fail.

Reply:
I think the resolve script already tests this.

I've created a dummy project exactly like the project set up in this test, just `s/no-dep/typescript/` because no-dep isn't a real package and typescript doesn't have dependencies.

```js
require('module').createRequire(require.resolve('workspace/package.json')).resolve('typescript')
```

throws with the current master build

```
> require('module').createRequire(require.resolve('workspace/package.json')).resolve('typescript')
Uncaught:
Error: Your application tried to access typescript, but it isn't declared in your dependencies; this makes the require call ambiguous and unsound.

Required package: typescript (via "typescript")
Required by: /private/var/folders/_d/ch2kc4h960d10cy_2c41qqzw0000gn/T/tmp.VtROGpaoX8/.yarn/$$virtual/workspace-virtual-4def450de0/1/workspace/

Require stack:
- /private/var/folders/_d/ch2kc4h960d10cy_2c41qqzw0000gn/T/tmp.VtROGpaoX8/.yarn/$$virtual/workspace-virtual-4def450de0/1/workspace/package.json
    at internalTools_makeError (/private/var/folders/_d/ch2kc4h960d10cy_2c41qqzw0000gn/T/tmp.VtROGpaoX8/.yarn/$$virtual/workspace-virtual-4def450de0/1/.pnp.js:3757:34)
    at resolveToUnqualified (/private/var/folders/_d/ch2kc4h960d10cy_2c41qqzw0000gn/T/tmp.VtROGpaoX8/.yarn/$$virtual/workspace-virtual-4def450de0/1/.pnp.js:4674:21)
    at resolveRequest (/private/var/folders/_d/ch2kc4h960d10cy_2c41qqzw0000gn/T/tmp.VtROGpaoX8/.yarn/$$virtual/workspace-virtual-4def450de0/1/.pnp.js:4766:29)
    at Object.resolveRequest (/private/var/folders/_d/ch2kc4h960d10cy_2c41qqzw0000gn/T/tmp.VtROGpaoX8/.yarn/$$virtual/workspace-virtual-4def450de0/1/.pnp.js:4832:26)
    at Function.external_module_.Module._resolveFilename (/private/var/folders/_d/ch2kc4h960d10cy_2c41qqzw0000gn/T/tmp.VtROGpaoX8/.pnp.js:3989:34)
    at Function.resolve (internal/modules/cjs/helpers.js:79:19)
> 
```

with the patch in the ticket (no time to pull this branch, build and test the actual code)

```
> require('module').createRequire(require.resolve('workspace/package.json')).resolve('typescript')
'/private/var/folders/_d/ch2kc4h960d10cy_2c41qqzw0000gn/T/tmp.VtROGpaoX8/.yarn/cache/typescript-patch-b4663f26e8-7e6169da2b.zip/node_modules/typescript/lib/typescript.js'
> 
```

Reply:
Hmm that's weird - the test doesn't crash by itself, but I do reproduce the crash in a regular environment 🤔

Reply:
Found it - there was a separate bug that was incorrectly causing the loose mode to be enabled when using `pnpFallbackMode: none` (which is the mode our tests use).